File: tests/application/GetOrderHandler.test.ts
// ============================================================
// Query Handler 測試 - 查詢訂單
// ============================================================
// 測試 CQRS 的查詢端，驗證讀取模型的正確性。
// ============================================================

import { CreateOrderHandler } from '../../src/application/commands/CreateOrderHandler';
import { GetOrderHandler, ListOrdersHandler } from '../../src/application/queries/GetOrderHandler';
import { InMemoryOrderCommandRepository } from '../../src/adapters/output/persistence/InMemoryOrderCommandRepository';
import { InMemoryOrderQueryRepository } from '../../src/adapters/output/persistence/InMemoryOrderQueryRepository';
import { InMemoryEventPublisher } from '../../src/adapters/output/messaging/InMemoryEventPublisher';
import { FixedIdGenerator } from '../../src/adapters/output/IdGeneratorAdapter';
import { OrderNotFoundError } from '../../src/domain/errors/DomainError';

describe('GetOrderHandler（CQRS 查詢端）', () => {
  let createHandler: CreateOrderHandler;
  let getOrderHandler: GetOrderHandler;
  let listOrdersHandler: ListOrdersHandler;
  let commandRepo: InMemoryOrderCommandRepository;
  let queryRepo: InMemoryOrderQueryRepository;
  let eventPublisher: InMemoryEventPublisher;

  beforeEach(() => {
    commandRepo = new InMemoryOrderCommandRepository();
    queryRepo = new InMemoryOrderQueryRepository();
    eventPublisher = new InMemoryEventPublisher(commandRepo, queryRepo);

    getOrderHandler = new GetOrderHandler(queryRepo);
    listOrdersHandler = new ListOrdersHandler(queryRepo);
  });

  it('應該能查詢已建立的訂單', async () => {
    const idGen = new FixedIdGenerator('order-q1');
    createHandler = new CreateOrderHandler(commandRepo, eventPublisher, idGen);

    await createHandler.execute({
      customerId: 'c1',
      items: [{ productId: 'p1', productName: '鍵盤', unitPrice: 2500, quantity: 1 }],
    });

    const result = await getOrderHandler.execute('order-q1');

    expect(result.id).toBe('order-q1');
    expect(result.customerId).toBe('c1');
    expect(result.totalAmount).toBe(2500);
    expect(result.items).toHaveLength(1);
    expect(result.items[0].productName).toBe('鍵盤');
  });

  it('查詢不存在的訂單應該拋出錯誤', async () => {
    await expect(getOrderHandler.execute('non-existent')).rejects.toThrow(OrderNotFoundError);
  });

  it('應該能列出所有訂單', async () => {
    const idGen = new FixedIdGenerator('order-1', 'order-2');
    createHandler = new CreateOrderHandler(commandRepo, eventPublisher, idGen);

    await createHandler.execute({
      customerId: 'c1',
      items: [{ productId: 'p1', productName: '鍵盤', unitPrice: 2500, quantity: 1 }],
    });
    await createHandler.execute({
      customerId: 'c2',
      items: [{ productId: 'p2', productName: '滑鼠', unitPrice: 800, quantity: 1 }],
    });

    const result = await listOrdersHandler.execute();
    expect(result).toHaveLength(2);
  });

  it('應該能依客戶 ID 篩選訂單', async () => {
    const idGen = new FixedIdGenerator('order-1', 'order-2', 'order-3');
    createHandler = new CreateOrderHandler(commandRepo, eventPublisher, idGen);

    await createHandler.execute({
      customerId: 'c1',
      items: [{ productId: 'p1', productName: '鍵盤', unitPrice: 2500, quantity: 1 }],
    });
    await createHandler.execute({
      customerId: 'c2',
      items: [{ productId: 'p2', productName: '滑鼠', unitPrice: 800, quantity: 1 }],
    });
    await createHandler.execute({
      customerId: 'c1',
      items: [{ productId: 'p3', productName: '螢幕', unitPrice: 12000, quantity: 1 }],
    });

    const result = await listOrdersHandler.execute('c1');
    expect(result).toHaveLength(2);
    expect(result.every((o) => o.customerId === 'c1')).toBe(true);
  });
});

File: src/test/java/com/example/hexagonal/application/query/GetOrderHandlerTest.java
package com.example.hexagonal.application.query;

import com.example.hexagonal.adapter.output.FixedIdGenerator;
import com.example.hexagonal.adapter.output.messaging.InMemoryEventPublisher;
import com.example.hexagonal.adapter.output.persistence.InMemoryOrderCommandRepository;
import com.example.hexagonal.adapter.output.persistence.InMemoryOrderQueryRepository;
import com.example.hexagonal.application.command.CreateOrderHandler;
import com.example.hexagonal.application.dto.CreateOrderInput;
import com.example.hexagonal.application.dto.OrderItemInput;
import com.example.hexagonal.domain.exception.OrderNotFoundException;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import java.util.List;

import static org.assertj.core.api.Assertions.*;

@DisplayName("GetOrderHandler / ListOrdersHandler（CQRS 查詢端）")
class GetOrderHandlerTest {

    private InMemoryOrderQueryRepository queryRepo;
    private GetOrderHandler getHandler;
    private ListOrdersHandler listHandler;
    private CreateOrderHandler createHandler;

    @BeforeEach
    void setUp() {
        var commandRepo = new InMemoryOrderCommandRepository();
        queryRepo = new InMemoryOrderQueryRepository();
        var eventPublisher = new InMemoryEventPublisher(commandRepo, queryRepo);
        getHandler = new GetOrderHandler(queryRepo);
        listHandler = new ListOrdersHandler(queryRepo);
        createHandler = new CreateOrderHandler(commandRepo, eventPublisher,
                new FixedIdGenerator("order-1", "order-2"));
    }

    @Test
    void 應該能查詢已建立的訂單() {
        createHandler.execute(new CreateOrderInput("c1", List.of(
                new OrderItemInput("p1", "鍵盤", 2500, 1))));

        var view = getHandler.execute("order-1");
        assertThat(view.id()).isEqualTo("order-1");
        assertThat(view.totalAmount()).isEqualTo(2500);
    }

    @Test
    void 查詢不存在的訂單應該拋出錯誤() {
        assertThatThrownBy(() -> getHandler.execute("non-existent"))
                .isInstanceOf(OrderNotFoundException.class);
    }

    @Test
    void 應該能列出所有訂單() {
        createHandler.execute(new CreateOrderInput("c1", List.of(
                new OrderItemInput("p1", "鍵盤", 2500, 1))));
        createHandler.execute(new CreateOrderInput("c2", List.of(
                new OrderItemInput("p2", "滑鼠", 800, 1))));

        var all = listHandler.execute(null);
        assertThat(all).hasSize(2);
    }

    @Test
    void 應該能依客戶ID篩選訂單() {
        createHandler.execute(new CreateOrderInput("c1", List.of(
                new OrderItemInput("p1", "鍵盤", 2500, 1))));
        createHandler.execute(new CreateOrderInput("c2", List.of(
                new OrderItemInput("p2", "滑鼠", 800, 1))));

        var filtered = listHandler.execute("c1");
        assertThat(filtered).hasSize(1);
        assertThat(filtered.getFirst().customerId()).isEqualTo("c1");
    }
}
